src/io.rs

```diff
@@ -33,3 +35,135 @@ impl std::error::Error for EspIOError {}
 pub mod asynch {
     pub use embedded_io_async::*;
 }
+
+/// A driver for bufferd / blocking communication ontop of Standard IO Streams.
+///
+/// By default all communication via std::io:stdin / stdout is non-blocking.
+/// That means if a user wants to read from stdin, they have to constantly poll the driver, since
+/// the respective hardware FIFO buffers are relative small. Also users would have to handle WouldBlock
+/// errors on every call, thouse make it unergonomic to handle.
+///
+/// A BlockingIoDriver will instruct the VFS(Virtual File System) to use the drivers interrupt driven,
+/// blocking read and write functions instead.
+pub struct BlockingIoDriver<T> {
+    phantom: PhantomData<T>,
+}
+
+pub struct BlockingUartIo<T: Uart> {
+    _driver: BlockingIoDriver<T>,
+}
+
+#[cfg(esp_idf_soc_usb_serial_jtag_supported)]
+use crate::usb_serial::UsbSerial;
+#[cfg(esp_idf_soc_usb_serial_jtag_supported)]
+pub struct BlockingSerialIo<T: UsbSerial> {
+    _driver: BlockingIoDriver<T>,
+}
+
+use crate::sys::esp;
+use crate::sys::{
+    esp_vfs_dev_uart_use_driver, esp_vfs_dev_uart_use_nonblocking, uart_driver_delete,
+    uart_driver_install, uart_is_driver_installed,
+};
+use crate::uart::Uart;
+
+impl<T: Uart> BlockingIoDriver<T> {
```

Comment:
Putting aside whether we should use an initialization "token" (a `BlockingIoDriver` instance) in the first place, as opposed to a simple function - it is not correct that we expose and allow both `BlockingIoDriver::uart` and `BlockingIoDriver::serial` to exist simultaneously and unconditionally:
* Only one of these should actually exist. This depends on whether the user has selected uart, jtag or (in future) cdc for stdin/out/err redirection at compile-time
* Therefore, it can be **a single API** named just `new` - or - for the case of a simple global function - `set_stdio_mode` without talking about / revealing details if the underlying peripheral is uart, jtag or cdc - user does not need to know or care, as all three support switching between blocking and nonblocking mode
* Ditto for the secondary console - we need a similar peripheral-agnostic API
* This API should be conditionalized and not even exist if user has selected `CONFIG_..._NONE` for the primary or secondary console. And in that case all peripherals should of course be available in the `Peripherals` struct from the get-go